src/session_server_ssh.c

```diff
@@ -468,20 +468,167 @@ nc_server_ssh_get_system_keys(const char *username, struct nc_public_key **pubke
     return ret;
 }
 
+#ifdef HAVE_SHADOW
+
+/**
+ * @brief Get the user's /etc/passwd entry.
+ *
+ * @param[in] username Username.
+ * @param[out] pwd_buf Buffer for the passwd structure.
+ * @param[out] buf Buffer for the pwd's strings.
+ * @param[out] buf_size Size of the buffer.
+ * @return User's passwd entry or NULL on error.
+ */
+static struct passwd *
+nc_server_ssh_getpwnam(const char *username, struct passwd *pwd_buf, char **buf, size_t *buf_size)
+{
+    struct passwd *pwd = NULL;
+    char *mem;
+    int r = 0;
+
+    do {
+        r = getpwnam_r(username, pwd_buf, *buf, *buf_size, &pwd);
+        if (pwd) {
+            /* entry found */
+            break;
+        }
+
+        if (r == ERANGE) {
+            /* small buffer, enlarge */
+            *buf_size <<= 2;
+            mem = realloc(*buf, *buf_size);
+            if (!mem) {
+                ERRMEM;
+                return NULL;
+            }
+            *buf = mem;
+        }
+    } while (r == ERANGE);
+
+    return pwd;
+}
+
+/**
+ * @brief Get the user's /etc/shadow entry.
+ *
+ * @param[in] username Username.
+ * @param[out] spwd_buf Buffer for the spwd structure.
+ * @param[out] buf Buffer for the spwd's strings.
+ * @param[out] buf_size Size of the buffer.
+ * @return User's shadow entry or NULL on error.
+ */
+static struct spwd *
+nc_server_ssh_getspnam(const char *username, struct spwd *spwd_buf, char **buf, size_t *buf_size)
+{
+    struct spwd *spwd = NULL;
+    char *mem;
+    int r = 0;
+
+    do {
+# ifndef __QNXNTO__
+        r = getspnam_r(username, spwd_buf, *buf, *buf_size, &spwd);
+# else
+        spwd = getspnam_r(username, spwd_buf, *buf, *buf_size);
+# endif
+        if (spwd) {
+            /* entry found */
+            break;
+        }
+
+        if (r == ERANGE) {
+            /* small buffer, enlarge */
+            *buf_size <<= 2;
+            mem = realloc(*buf, *buf_size);
+            if (!mem) {
+                ERRMEM;
+                return NULL;
+            }
+            *buf = mem;
+        }
+    } while (r == ERANGE);
+
+    return spwd;
+}
+
+/**
+ * @brief Get the user's hashed password from the system.
+ *
+ * @param[in] username Username.
+ * @return User's hashed password or NULL on error.
+ */
+static char *
+nc_server_ssh_get_pwd_hash(const char *username)
+{
+    struct passwd *pwd, pwd_buf;
+    struct spwd *spwd, spwd_buf;
+    char *pass_hash = NULL, *buf = NULL;
+    size_t buf_size = 256;
+
+    buf = malloc(buf_size);
+    NC_CHECK_ERRMEM_GOTO(!buf, , error);
+
+    pwd = nc_server_ssh_getpwnam(username, &pwd_buf, &buf, &buf_size);
+    if (!pwd) {
+        VRB(NULL, "User \"%s\" not found locally.", username);
```

Comment:
The YANG-level feature which enables this behavior is called `local-users-supported`, and if it's enabled, then "local" refers to YANG config. Please change this string so that it says "in the system configuration" or "outside of YANG data" or something, but let's not call them "local".

Also everywhere else, e.g., the docstring for `nc_server_ssh_auth_kbdint_system` and the local variables in there.

Reply:
Right, changed these occurrences to contain the word `system`.